### PR TITLE
GAWB-140: Persist WorkflowFailure when failing to submit to Cromwell

### PR DESCRIPTION
**Pull request by abaumann**
Modified it so we do not retry submissions with cromwell on 500 error codes, and instead convert those failures to submit into WorkflowFailure(s).  
Added unit test to make sure WorkflowFailure was persisted into the DB on a timeout.  
MockExecutionServiceDAO supports faked timeouts now.
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala [hunk 1]
```diff
@@ -23,8 +23,7 @@ class HttpExecutionServiceDAO( executionServiceURL: String )( implicit val syste
     import system.dispatcher
     val pipeline = addAuthHeader(userInfo) ~> sendReceive ~> unmarshal[ExecutionServiceStatus]
     val formData = FormData(Seq("wdlSource" -> wdl, "workflowInputs" -> inputs) ++ options.map("workflowOptions" -> _).toSeq)
-    retry(when500) { () => pipeline(Post(url,formData)) }
-//    retry(Await.result(pipeline(Post(url,formData)),Duration.Inf),when500)
+    pipeline(Post(url,formData))
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala:26]: If we get a 5xx, but Cromwell actually submits the workflow, we don't want to actually retry and end up with orphaned workflows

File: src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala [hunk 2]
```diff
@@ -319,6 +323,24 @@ class SubmissionSpec(_system: ActorSystem) extends TestKit(_system) with FlatSpe
...
+    val submissionStatusRq = Await.result(workspaceService.getSubmissionStatus(testData.wsName, newSubmission.submissionId), Duration.Inf).asInstanceOf[RequestComplete[(StatusCode, SubmissionStatusResponse)]]
+    val (submissionStatus, submissionStatusResponse) = submissionStatusRq.response
+    assertResult(StatusCodes.OK) {
+      submissionStatus
+    }
+
+    assert(submissionStatusResponse.notstarted.size == 1)
+    assert(submissionStatusResponse.workflows.size == 0)
+  }
+
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala]: can you add a test here to check that you correctly concat normal-failures and timeout-failures together

(yes i know this is annoying, sorry)

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -960,14 +960,24 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
...
-          execStatusFuture map (execStatus => Workflow(workflowId = execStatus.id, status = WorkflowStatuses.Submitted, statusLastChangedDate = DateTime.now, workflowEntity = Option(AttributeEntityReference(entityType = header.entityType, entityName = entityInputs.entityName)), inputResolutions = entityInputs.inputResolutions))
+          execStatusFuture map {
+            execStatus => Workflow(workflowId = execStatus.id, status = WorkflowStatuses.Submitted, statusLastChangedDate = DateTime.now, workflowEntity = Option(AttributeEntityReference(entityType = header.entityType, entityName = entityInputs.entityName)), inputResolutions = entityInputs.inputResolutions)
+          } recover {
+            case t: Exception => WorkflowFailure(entityInputs.entityName, header.entityType, entityInputs.inputResolutions, Seq(AttributeString("Unable to submit workflow")))
+          }
         }))
 
         submittedWorkflowsFuture map { submittedWorkflows =>
+          val succeededWorkflowSubmissions = submittedWorkflows.collect {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:985]: it's tragic that `collect` is the accepted way to filter on class :(

Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:985]: Not sure if this is the right way to do it, looking for feedback

File: src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala (hunk 2 above)
Comment (abaumann) [inline src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala]: no, that's a good suggestion, will do

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 3 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:985]: I asked Google and it seems to be.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 4]
```diff
@@ -960,14 +960,24 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
...
           val execStatusFuture = executionServiceDAO.submitWorkflow(wdl, MethodConfigResolver.propertiesToWdlInputs(methodProps.toMap), workflowOptions, userInfo)
-          execStatusFuture map (execStatus => Workflow(workflowId = execStatus.id, status = WorkflowStatuses.Submitted, statusLastChangedDate = DateTime.now, workflowEntity = Option(AttributeEntityReference(entityType = header.entityType, entityName = entityInputs.entityName)), inputResolutions = entityInputs.inputResolutions))
+          execStatusFuture map {
+            execStatus => Workflow(workflowId = execStatus.id, status = WorkflowStatuses.Submitted, statusLastChangedDate = DateTime.now, workflowEntity = Option(AttributeEntityReference(entityType = header.entityType, entityName = entityInputs.entityName)), inputResolutions = entityInputs.inputResolutions)
+          } recover {
+            case t: Exception => WorkflowFailure(entityInputs.entityName, header.entityType, entityInputs.inputResolutions, Seq(AttributeString("Unable to submit workflow")))
+          }
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: This works in the case where execStatusFuture fails but what happens if there is an exception in either of the 2 preceding lines? I suggest you also put a try around the whole block and if there is an exception return
Future.successful(WorkflowFailure(...

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 5]
```diff
@@ -958,16 +959,38 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
...
+                val error = "Unable to submit workflow when creating submission"
+                logger.error(error, t)
+                WorkflowFailure(entityInputs.entityName, header.entityType, entityInputs.inputResolutions, Seq(AttributeString(error), AttributeString(t.getMessage)))
+              }
+            }
+          } match {
+            case Success(result) => result
+            case Failure(t) => {
+              val error = "Unable to process workflow when creating submission"
+              Future.successful(WorkflowFailure(entityInputs.entityName, header.entityType, entityInputs.inputResolutions, Seq(AttributeString(error), AttributeString(t.getMessage))))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: should log this too

**Comment by dvoet** [comment]
add 1 log line then :+1:

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 5 above)
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: meant to do that, good catch

**Pull request merged**